File: cross_correlation.js
/*
* Author: Chris Garry
* Description:
*	Implements cross-correlation of series x and y as defined by
*
*							sigma_i[(x(i)-mean_x)*(y(i-d)-mean_y)]
*		r(d) =	-----------------------------------------------------------------
*				sqrt(sigma_i[(x(i)-mean_x)^2]) * sqrt(sigma_i[(y(i-d)-mean_y)^2])
*
* References: 
* 	http://paulbourke.net/miscellaneous/correlate/
*/


function InputException(message) {
   this.message = message;
   this.name = "InputException";
}

var cross_correlation=function(series_x, series_y) 
{	
	var sum_x; 
	var sum_y; 
	var mean_x;
	var mean_y;
	var length_x = series_x.length;
	var length_y = series_y.length;
	var sqr_series_x;
	var sqr_series_y;
	var cross_correlate = new Array();
	var delay_max = 2*length_x;

	//If time series of valid length.
		if((length_x!=0)&&(length_y!=0){
			
			//If series of different lengths, pad with zeros.
				var diff = length_x-length_y;

				if(diff>0){
					for(int x=0; x<diff; x++){
						series_y[length_y+x] = 0;
					}
				}

				if(diff<0){
					for(int x=0; x<diff; x++){
						series_x[length_x+x] = 0;
					}
					//Output size is relative to larger array of x,y
					delay_max = 2*length_y;
				}

			//Calculate sum of each series.
				for(int z = 0; z<length_x; z++){
					sum_x += series_x[z];
					sum_y += series_y[z];
				}

			//Calculate mean of each series.
				mean_x = sum_x/length_x;
				mean_y = sum_y/length_y;

			//Compute cross-correlate coefficients
				for(int delay = -delay_max; delay<delay_max; delay++){

					numerator = 0;
					sqr_series_x = 0;
					sqr_series_y = 0;

					for(int i =0; i<length_x; i++){

						//Skip cofficients with out of bound indicies
						var j = i + delay;
						if (j < 0 || j >= length_x){
							continue;
						}
						else{

						var term_1 = (series_x[i]-mean_x);
						var term_2 = (series_y[j]-mean_y);	
							numerator 	 += term_1*term_2;
							sqr_series_x += term_1*term_1;
							sqr_series_y += term_2*term_2;
						}

					}

					cross_correlate[delay] = numerator/Math.sqrt(sqr_series_x*sqr_series_y);
				}

			return cross_correlate;
	  	}

	//Invalid time series input
	  	else{
	  		throw new InputException("Inputs must be of non-zero length.");
	  	}
}

File: cross_correlation.c
/*
	Author: Chris Garry, chrislgarry AT gmail DOT com
	
	Description:
		The following code implements a handshake matching algorithm
    using cross correlation given accelerometer data from a Pebble Smart
		Watch. 

	License:

		This code is for education purposes only.

		Reference for cross_correlation():
			Paul Bourke, August 1996, Cross Correlation, AutoCorrelation -- 2D Pattern Identification 
			http://paulbourke.net/miscellaneous/correlate/
*/

#include "handshake.h"

//HELPER FUNCTIONS==========================================================================
	
	//Computes the cross correlation coefficients for series_x and series_y
	double *cross_correlation(int *series_x, int *series_y){	

		double sum_x = 0.0;
		double sum_y = 0.0;
		double mean_x = 0.0;
		double mean_y = 0.0;
		int delay_max = SERIES_X_SIZE;
		double *cross_correlate = (double*)malloc(((delay_max*2)+1)*sizeof(double));

		//Calculate the sum of each series
			int z;
			for(z = 0; z<SERIES_X_SIZE; z++){
				sum_x += series_x[z];
				sum_y += series_y[z];
			}

		//Calculate mean of each series.
			mean_x = sum_x/SERIES_X_SIZE;
			mean_y = sum_y/SERIES_Y_SIZE;

		//Compute cross-correlate coefficients
			double term_1, term_2, numerator;
			double sqr_series_x = 0.0;
			double sqr_series_y = 0.0;
			int delay, i;
			for(delay = -delay_max; delay<=delay_max; delay++){

				numerator = 0;
				sqr_series_x = 0;
				sqr_series_y = 0;

				for(i =0; i<SERIES_X_SIZE; i++){

					//Assume cofficients with out of bound indicies have x[i]=0, y[i]=0
					int j = i - delay;
					if (j < 0 || j >= SERIES_X_SIZE){
						term_1 = (series_x[i]-mean_x);
						term_2 = (-1*mean_y);	
						numerator += term_1*term_2;
						sqr_series_x += term_1*term_1;
						sqr_series_y += term_2*term_2;			
					}
					else{
						term_1 = (series_x[i]-mean_x);
						term_2 = (series_y[j]-mean_y);	
						numerator += term_1*term_2;
						sqr_series_x += term_1*term_1;
						sqr_series_y += term_2*term_2;
					}

				}

				cross_correlate[delay+delay_max] = numerator/sqrt(sqr_series_x*sqr_series_y);
				printf("delay:%d correlate:%f\n",delay, cross_correlate[delay+delay_max]);
			}

		return cross_correlate;
	}
    
//END HELPER FUNCTIONS======================================================================


//LIBRARY FUNCTIONS=========================================================================
	int handshake_match_xcorr(int *series_x, int *series_y){

		//Compute the cross correlation coefficients
			double *coefficients = cross_correlation(series_x,series_y);

		//Determine if a correlation coefficient >XCORR_THRESH exists.
			int i;
			for(i=0; i<SERIES_X_SIZE*2; i++){
				if(fabs(coefficients[i]) > XCORR_THRESH){
					return MATCH;
				}
			}

		return NO_MATCH;
	}
//END LIBRARY FUNCTIONS=====================================================================


//TEST======================================================================================
	int main(){

		//TEST XCORR==============================================================
			int test_xcorr_1[14] = {-1020,-1774,-972,34,294,-498,-1796,-1034,432,-1144,-1892,-1482,-508,-1176};
			int test_xcorr_2[14] = {-1014,-1166,-920,64,148,-734,-1024,-1012,-1856,366,-1842,-1330,-108,-1746};

		    if(handshake_match_xcorr(test_xcorr_1, test_xcorr_2)){
		    	printf("\nXCORR RESULT: MATCH.\n");
		    }
		    else{
		    	printf("\nXCORR RESULT: NO MATCH.\n");
		    }

		return 0;
	 }
